get dagster-mlflow passing dagster build

sryza opened the pull request:
This pull request gets dagster-mlflow passing dagster's build and implements some of the suggestions that I made in the base pull request.

The changes bundled into this pull request:
* Fixes lint complaints and import-ordering complaints.
* Consolidates the hooks into a single `end_mlflow_run_on_pipeline_finished ` hook.
* Adds a test that uses the resource inside a pipeline.
* Suggests usage of "mlflow.log_params" instead of "context.resources.mlflow.log_params".

sryza committed:
mlflow CI

Summary:
fixed init file

added tests

fixed tests

tidying up tests

updated docstring with more detailed usage instructions

updated success hook and test for it

mlflow updates

Test Plan: bk

Differential Revision: https://dagster.phacility.com/D8335

hug0l1ma commented on python_modules/libraries/dagster-mlflow/dagster_mlflow/resources.py:
```diff
@@ -225,74 +224,32 @@ def chunks(params: dict, size: int = 100):
-
-                mlf.log_params(some_params)
-                mlf.tracking_client.create_registered_model(some_model_name)
+                mlflow.log_params(some_params)
```
So this does work to just use the mlflow api directly here instead of the resource and everything will be logged to the correct mlflow run even with multiple processes. For the specific case of log_params we override the method in the resource as it raises an error if `len(some_params)>100`, so I would maybe encourage using the method from the resource. I guess the reason I was encouraging the use of the resource rather than the mlflow api directly is because in our use case we also subclass the resource class to add some more methods that are particularly useful for us (and we may contribute them in the future if we think they are useful for others and general enough). Here I guess it doesn't make too much of a difference though.

hug0l1ma commented on python_modules/libraries/dagster-mlflow/dagster_mlflow/resources.py:
```diff
@@ -225,74 +224,32 @@ def chunks(params: dict, size: int = 100):
-                mlf.log_params(some_params)
-                mlf.tracking_client.create_registered_model(some_model_name)
+                mlflow.log_params(some_params)
+                mlflow.tracking_client.create_registered_model(some_model_name)
```
This wouldn't work if we're using the mlflow module directly since this is not defined. In the resource we add the tracking client as an attribute of the resource and so could use it this way. To use the tracking client from the mlflow module api you would need to use `mlflow.tracking.MlflowClient()`.

hug0l1ma commented on python_modules/libraries/dagster-mlflow/dagster_mlflow_tests/test_resources.py:
```diff
@@ -260,117 +263,131 @@ def test_setup(mock_atexit, context):
+
+    result = execute_pipeline(
+        mlf_pipeline,
+        run_config={
```
might be worth using the multiprocess executor for this test as that would be the case when we would expect the mlflow run ids to be different if the mlflow resource is not used

hug0l1ma reviewed (changes requested):
This looks really good. Thanks for making the changes @sryza and sorry I didn't have time to do them myself earlier. Just some minor comments regarding the resource docstring and the new test.

sryza commented:
@hug0l1ma just an update: I thought your suggestion about using the multiprocess executor for the test was a great one.  I wrote one and have been getting some strange errors, that I'm in the process of debugging.

sryza committed:
fix tracking client

sryza commented:
I suspect that this or something similar to it is causing the issue I was experiencing when trying to write a multiprocessing test: https://issues.apache.org/jira/browse/SPARK-22674.  Have you gotten mlflow to work with Dagster multiprocessing?

I'm not sure how to work around that, so I reverted to single processing.  I fixed the tracking client issue that you brought up.

hug0l1ma commented:
> I suspect that this or something similar to it is causing the issue I was experiencing when trying to write a multiprocessing test: https://issues.apache.org/jira/browse/SPARK-22674. Have you gotten mlflow to work with Dagster multiprocessing?
> 
> I'm not sure how to work around that, so I reverted to single processing. I fixed the tracking client issue that you brought up.

Hi @sryza, I do have mlflow working with Dagster multiprocessing (also tested running on a cluster with the CeleryK8sExecutor and it logs to a single mlflow run as expected) so I'm not too worried that there is an issue writing the test for it. I'll go ahead and merge into my branch and then I guess we're good to merge to master? Thanks again for making all the changes and looking into this, it would have taken longer for me to find the time to do it.
